File: src/Catty/Extension&Delegate&Protocol/Extensions/UIColor+Extension/UIColor+CatrobatUIColorExtensions.m
/**
 *  Copyright (C) 2010-2018 The Catrobat Team
 *  (http://developer.catrobat.org/credits)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  An additional term exception under section 7 of the GNU Affero
 *  General Public License, version 3, is available at
 *  (http://developer.catrobat.org/license_additional_term)
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 */

#import "UIColor+CatrobatUIColorExtensions.h"

@implementation UIColor (CatrobatUIColorExtensions)

// taken from https://github.com/anjerodesu/UIColor-ColorWithHex/blob/master/UIColor%2BColorWithHex.m
+ (UIColor *)colorWithHex:(UInt32)hexadecimal
{
	CGFloat red, green, blue;

	// bitwise AND operation
	// hexadecimal's first 2 values
	red = ( hexadecimal >> 16 ) & 0xFF;
	// hexadecimal's 2 middle values
	green = ( hexadecimal >> 8 ) & 0xFF;
	// hexadecimal's last 2 values
	blue = hexadecimal & 0xFF;

	UIColor *color = [UIColor colorWithRed: red / 255.0f green: green / 255.0f blue: blue / 255.0f alpha: 1.0f];
	return color;
}
#pragma mark intern Colors

+ (UIColor*)lightColor
{
    return [UIColor colorWithHex:0xADEEF0];
}

+ (UIColor*)mediumColor
{
    return [UIColor colorWithHex:0x18A5B7];
}

+ (UIColor*)darkColor
{
    return [UIColor colorWithHex:0x191919];
}

+(UIColor*)whiteGrayColor
{
    return [UIColor colorWithRed:250.0f/255.0f green:250.0f/255.0f blue:250.0f/255.0f alpha:1.0f];
}

+(UIColor*)textViewBorderGrayColor
{
    return [UIColor colorWithRed:225.0f/255.0f green:225.0f/255.0f blue:225.0f/255.0f alpha:1.0f];
}

+(UIColor*)destructiveColor
{
    return [UIColor colorWithHex: 0xF26C4F];
}

# pragma mark Global

+ (UIColor*)globalTintColor
{
    return [self mediumColor];
}

+ (UIColor*)utilityTintColor
{
    return [self mediumColor];
}

+ (UIColor*)navBarColor
{
    return [self mediumColor];
}

+ (UIColor*)navTintColor
{
    return [self lightColor];
}

+ (UIColor*)navTextColor
{
    return [self backgroundColor];
}

+ (UIColor*) toolBarColor
{
    return [self navBarColor];
}

+ (UIColor*) toolTintColor
{
    return [self navTintColor];
}

+ (UIColor*) tabBarColor
{
    return [self navBarColor];
}

+ (UIColor*) tabTintColor
{
    return [self navTintColor];
}

+ (UIColor*)buttonTintColor
{
    return [self mediumColor];
}

+ (UIColor*)textTintColor
{
    return [self darkColor];
}

+ (UIColor*)buttonHighlightedTintColor
{
    return [self backgroundColor];
}

+ (UIColor*)destructiveTintColor
{
    return [self destructiveColor];
}

+ (UIColor*)backgroundColor
{
    return [self whiteColor];
}

# pragma mark FormulaEditor
+ (UIColor*)formulaEditorOperatorColor
{
    return [self buttonTintColor];
}
+ (UIColor*)formulaEditorBorderColor
{
    return [self lightColor];
}

+ (UIColor*)formulaButtonTextColor
{
    return [self lightColor];
}

+ (UIColor*)formulaEditorHighlightColor
{
    return [self buttonTintColor];
}

+ (UIColor*)formulaEditorOperandColor
{
    return [self buttonTintColor];
}

# pragma mark IDE
+ (UIColor*)brickSelectionBackgroundColor
{
    return [UIColor colorWithRed:13.0f/255.0f green:13.0f/255.0f blue:13.0f/255.0f alpha:1.0f];
}

+ (UIColor*)lookBrickGreenColor
{
    return [UIColor colorWithRed:57.0f/255.0f green:171.0f/255.0f blue:45.0f/255.0f alpha:1.0f];
}

+ (UIColor*)lookBrickStrokeColor
{
    return [UIColor colorWithRed:185.0f/255.0f green:220.0f/255.0f blue:110.0f/255.0f alpha:1.0f];
}

+ (UIColor*)motionBrickBlueColor
{
    return [UIColor colorWithRed:29.0f/255.0f green:132.0f/255.0f blue:217.0f/255.0f alpha:1.0f];
}

+ (UIColor*)motionBrickStrokeColor
{
    return [UIColor colorWithRed:179.0f/255.0f green:203.0f/255.0f blue:255.0f/255.0f alpha:1.0f];
}

+ (UIColor*)controlBrickOrangeColor
{
    return [UIColor colorWithRed:255.0f/255.0f green:120.0f/255.0f blue:20.0f/255.0f alpha:1.0f];
}

+ (UIColor*)controlBrickStrokeColor
{
    return [UIColor colorWithRed:247.0f/255.0f green:208.0f/255.0f blue:187.0f/255.0f alpha:1.0f];
}

+ (UIColor*)varibaleBrickRedColor
{
    return [UIColor colorWithRed:234.0f/255.0f green:59.0f/255.0f blue:59.0f/255.0f alpha:1.0f];
}

+ (UIColor*)variableBrickStrokeColor
{
    return [UIColor colorWithRed:238.0f/255.0f green:149.0f/255.0f blue:149.0f/255.0f alpha:1.0f];
}

+ (UIColor*)soundBrickVioletColor
{
    return [UIColor colorWithRed:180.0f/255.0f green:67.0f/255.0f blue:198.0f/255.0f alpha:1.0f];
}

+ (UIColor*)soundBrickStrokeColor
{
    return [UIColor colorWithRed:179.0f/255.0f green:137.0f/255.0f blue:255.0f/255.0f alpha:1.0f];
}

+ (UIColor*)PhiroBrickColor
{
    return [UIColor colorWithRed:234.0f/255.0f green:200.0f/255.0f blue:59.0f/255.0f alpha:1.0f];
}

+ (UIColor*)PhiroBrickStrokeColor
{
    return [UIColor colorWithRed:179.0f/255.0f green:137.0f/255.0f blue:255.0f/255.0f alpha:1.0f];
}

+ (UIColor*)ArduinoBrickColor
{
    return [UIColor colorWithRed:234.0f/255.0f green:200.0f/255.0f blue:59.0f/255.0f alpha:1.0f];
}

+ (UIColor*)ArduinoBrickStrokeColor
{
    return [UIColor colorWithRed:179.0f/255.0f green:137.0f/255.0f blue:255.0f/255.0f alpha:1.0f];
}

@end

File: src/Catty/Extension&Delegate&Protocol/Extensions/UIColor+Extension/UIColor+CatrobatUIColorExtensions.swift
/**
 *  Copyright (C) 2010-2020 The Catrobat Team
 *  (http://developer.catrobat.org/credits)
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  An additional term exception under section 7 of the GNU Affero
 *  General Public License, version 3, is available at
 *  (http://developer.catrobat.org/license_additional_term)
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see http://www.gnu.org/licenses/.
 */

@objc
extension UIColor {

    /* WORK IN PROGRESS: Theming
    @nonobjc
    static let currentTheme = CattyTheme.teal

    enum CattyTheme {
        case teal //default
        case red
        case pink
    }*/

    /*
     * from: https://stackoverflow.com/a/24263296
     */
    convenience init(red: Int, green: Int, blue: Int) {
        assert(red >= 0 && red <= 255, "Invalid red component")
        assert(green >= 0 && green <= 255, "Invalid green component")
        assert(blue >= 0 && blue <= 255, "Invalid blue component")

        self.init(red: CGFloat(red) / 255.0, green: CGFloat(green) / 255.0, blue: CGFloat(blue) / 255.0, alpha: 1.0)
    }

    convenience init(hex: Int) {
        self.init(
            red: (hex >> 16) & 0xFF,
            green: (hex >> 8) & 0xFF,
            blue: hex & 0xFF
        )
    }

    // MARK: intern Colors

    static var light: UIColor {
        /* WORK IN PROGRESS: Theming
        switch currentTheme {
        case .red:
            return UIColor(hex: 0xde746a)
        case .pink:
            return UIColor(hex: 0xe68cd8)
        default: //teal
            return UIColor(hex: 0xadeef0)
        }*/
        return UIColor(hex: 0xadeef0)
    }

    static var medium: UIColor {
        /* WORK IN PROGRESS: Theming
        switch currentTheme {
        case .red:
            return UIColor(hex: 0xad2215)
        case .pink:
            return UIColor(hex: 0xb7189f)
        default: //teal
            return UIColor(hex: 0x18a5b7)
        }*/
        return UIColor(hex: 0x18a5b7)
    }

    static var dark: UIColor {
        UIColor(hex: 0x191919)
    }

    static var whiteGray: UIColor {
        UIColor(red: 250.0 / 255.0, green: 250.0 / 255.0, blue: 250.0 / 255.0, alpha: 1.0)
    }

    static var textViewBorderGray: UIColor {
        UIColor(red: 225.0 / 255.0, green: 225.0 / 255.0, blue: 225.0 / 255.0, alpha: 1.0)
    }

    static var destructive: UIColor {
        UIColor(hex: 0xf26c4f)
    }

    // MARK: Global

    static var globalTint: UIColor {
        self.medium
    }

    static var utilityTint: UIColor {
        self.medium
    }

    static var navBar: UIColor {
        self.medium
    }

    static var navTint: UIColor {
        /* WORK IN PROGRESS: Dark mode
        if #available(iOS 13.0, *) {
            if UITraitCollection.current.userInterfaceStyle == .dark {
                return self.dark
            }
        }*/
        return self.light
    }

    static var navText: UIColor {
        self.background
    }

    static var toolBar: UIColor {
        self.navBar
    }

    static var toolTint: UIColor {
        self.navTint
    }

    static var tabBar: UIColor {
        self.navBar
    }

    static var tabTint: UIColor {
        self.navTint
    }

    static var buttonTint: UIColor {
        self.medium
    }

    static var textTint: UIColor {
        /* WORK IN PROGRESS: Dark mode
        if #available(iOS 13.0, *) {
            if UITraitCollection.current.userInterfaceStyle == .dark {
                return self.light
            }
        }*/
        return self.dark
    }

    static var pageIndicator: UIColor {
        UIColor(hex: 0x3ab2c1)
    }

    static var buttonHighlightedTint: UIColor {
        self.background
    }

    static var destructiveTint: UIColor {
        self.destructive
    }

    static var background: UIColor {
        if #available(iOS 13.0, *) {
            if UITraitCollection.current.userInterfaceStyle == .dark {
                return self.black
            }
        }
        return self.white
    }

    // MARK: FormulaEditor

    static var formulaEditorOperator: UIColor {
        self.buttonTint
    }

    static var formulaEditorHighlight: UIColor {
        self.buttonTint
    }

    static var formulaEditorOperand: UIColor {
        self.buttonTint
    }

    static var formulaEditorBorder: UIColor {
        self.light
    }

    static var formulaButtonText: UIColor {
        self.light
    }

    // MARK: IDE

    // Bricks & Scripts Colors
    static var brickSelectionBackground: UIColor {
        UIColor(red: 13.0 / 255.0, green: 13.0 / 255.0, blue: 13.0 / 255.0, alpha: 1.0)
    }

    static var lookBrickGreen: UIColor {
        UIColor(red: 57.0 / 255.0, green: 171.0 / 255.0, blue: 45.0 / 255.0, alpha: 1.0)
    }

    static var lookBrickStroke: UIColor {
        UIColor(red: 185.0 / 255.0, green: 220.0 / 255.0, blue: 110.0 / 255.0, alpha: 1.0)
    }

    static var motionBrickBlue: UIColor {
        UIColor(red: 29.0 / 255.0, green: 132.0 / 255.0, blue: 217.0 / 255.0, alpha: 1.0)
    }

    static var motionBrickStroke: UIColor {
        UIColor(red: 179.0 / 255.0, green: 203.0 / 255.0, blue: 255.0 / 255.0, alpha: 1.0)
    }

    static var controlBrickOrange: UIColor {
        UIColor(red: 255.0 / 255.0, green: 120.0 / 255.0, blue: 20.0 / 255.0, alpha: 1.0)
    }

    static var controlBrickStroke: UIColor {
        UIColor(red: 247.0 / 255.0, green: 208.0 / 255.0, blue: 187.0 / 255.0, alpha: 1.0)
    }

    static var variableBrickRed: UIColor {
        UIColor(red: 234.0 / 255.0, green: 59.0 / 255.0, blue: 59.0 / 255.0, alpha: 1.0)
    }

    static var variableBrickStroke: UIColor {
        UIColor(red: 238.0 / 255.0, green: 149.0 / 255.0, blue: 149.0 / 255.0, alpha: 1.0)
    }

    static var soundBrickViolet: UIColor {
        UIColor(red: 180.0 / 255.0, green: 67.0 / 255.0, blue: 198.0 / 255.0, alpha: 1.0)
    }

    static var soundBrickStroke: UIColor {
        UIColor(red: 179.0 / 255.0, green: 137.0 / 255.0, blue: 255.0 / 255.0, alpha: 1.0)
    }

    static var phiroBrick: UIColor {
        UIColor(red: 234.0 / 255.0, green: 200.0 / 255.0, blue: 59.0 / 255.0, alpha: 1.0)
    }

    static var phiroBrickStroke: UIColor {
        UIColor(red: 179.0 / 255.0, green: 137.0 / 255.0, blue: 255.0 / 255.0, alpha: 1.0)
    }

    static var arduinoBrick: UIColor {
        UIColor(red: 234.0 / 255.0, green: 200.0 / 255.0, blue: 59.0 / 255.0, alpha: 1.0)
    }

    static var arduinoBrickStroke: UIColor {
        UIColor(red: 179.0 / 255.0, green: 137.0 / 255.0, blue: 255.0 / 255.0, alpha: 1.0)
    }
}
